frontend/src/app/modules/matches/matches-computations.js

```diff
@@ -68,3 +68,49 @@ export const computeWinRatio = (playerId, playerMatches) => {
   const wonMatchesCount = playerMatches.filter(match => didPlayerWin(playerId, match)).length
   return (playerMatches.length > 0) ? (wonMatchesCount / playerMatches.length) : 0
 }
+
+export const computeKingStreakDuration = (matchesLast, playersLast) => {
+  if (playersLast.length < 1) {
+    return null
+  }
+
+  const playersMap = playersLast.reduce((map, player) => {
+    map[player.id] = player.rating
+    return map
+  }, new Map())
+
+  const kingId = playersLast[0].id
+  let matchFound = false
+  for (const match of matchesLast) {
```

Comment:
looks a bit like black magic to me :D i guess idea is to go backwards until we find some match (moment in time) where the king became a king, right?

Reply:
maybe if you could a bit elaborate on those 3 steps... The way i see how it can be computed is smth like that:
1. King may become a such by playing a match or by somebody else playing a match (and losing rating)
2. I think we have a snapshot of players ratings in their matches
3. If King became a such by other losing some rating then it seems easy - just find a match where someone has a bigger rating then King's. I guess it's right? :man_shrugging: 
4. If King became a such by gaining some rating then we need somehow to realize that the previous King has lower rating... Previous King can be or not in the match - so I am afraid we need to compare somehow to all other players here (or at least the second biggest at that moment - has to be computed somehow) and understand if this won match is this deal-breaker match...

Probably first 3 I somehow see in the code below but definitely missing 4th point... maybe I am missing something in my conclusions?...

Reply:
in any case not a trivial algorithm definitely needs more docs/comments or maybe we should just solve it by adding something simple in the BE instead

Reply:
@petrk-salsita @novellizator any thoughts? :)

Reply:
I added some brief explanation of the algorithm and it here as well:
```
  First, we create a hash table of players and their current scores
  Then we iterate through matches from the latest, recompute scores of affected players and compare
  More preciselly:
    1. Recompute king's score in case he played the match, note whether he won the match
    2. Recompute scores of all (other) players in the match and test if any of them tops king
    3. Otherwise, compare his new score with scores of all other players,
    in case the king played the match and won (thus had smaller score before).
```
How you described it is basically the way it is implemented, looping over players affected by the match (your 3.) is 2. in the docs (lines 84-87) and looping over all players in case King won the match (your 4.) is 3. in the docs (lines 89-93)